v2/googlecloud-to-elasticsearch/src/test/java/com/google/cloud/teleport/v2/elasticsearch/transforms/EventMetadataTest.java

```diff
@@ -0,0 +1,113 @@
+
+        String inputMessage = "{\"message\":\"test message\", \"timestamp\":\"2021-08-11T19:55:54.834736Z\"}";
+
+        EventMetadata eventMetadata = EventMetadata.build(inputMessage, options);
+
+        JsonNode enrichedMessageAsJson = eventMetadata.getEnrichedMessageAsJsonNode();
+        String enrichedMessageAsString = eventMetadata.getEnrichedMessageAsString();
+
+        Assert.assertTrue(StringUtils.isNotBlank(enrichedMessageAsString));
+        Assert.assertEquals(inputMessage, enrichedMessageAsJson.get("message").textValue());
+        Assert.assertEquals("999.999.999", enrichedMessageAsJson.get("agent").get("version").textValue());
+        Assert.assertEquals(Dataset.AUDIT.getKey(), enrichedMessageAsJson.get("data_stream").get("dataset").textValue());
+        Assert.assertEquals("test-namespace", enrichedMessageAsJson.get("data_stream").get("namespace").textValue());
+        Assert.assertEquals(Dataset.AUDIT.getKey(), enrichedMessageAsJson.get("service").get("type").textValue());
+        Assert.assertEquals("2021-08-11T19:55:54.834736Z", enrichedMessageAsJson.get("@timestamp").textValue());
+    }
+
+    @Test
+    public void testEventMetadataAppend() throws IOException {
+        PubSubToElasticsearchOptions options =
+                TestPipeline.testingPipelineOptions().as(PubSubToElasticsearchOptions.class);
+
+        options.setDeadletterTable("test:dataset.table");
+        options.setElasticsearchUsername("test");
+        options.setElasticsearchPassword("test");
+        options.setDataset(Dataset.AUDIT);
+        options.setNamespace("test-namespace");
+
+        String inputMessage = Files.lines(
+                Paths.get(
+                        IS_WINDOWS
+                                ? INPUT_MESSAGE_FILE_PATH.substring(1)
+                                : INPUT_MESSAGE_FILE_PATH), StandardCharsets.UTF_8)
+                .collect(Collectors.joining());
+
+        EventMetadata eventMetadata = EventMetadata.build(inputMessage, options);
+
+        JsonNode enrichedMessageAsJson = eventMetadata.getEnrichedMessageAsJsonNode();
+
+        //if elasticsearchTemplateVersion is not set, 1.0.0 is the default value
+        Assert.assertEquals("1.0.0", enrichedMessageAsJson.get("agent").get("version").textValue());
+        Assert.assertEquals(enrichedMessageAsJson.get("data_stream").get("dataset").textValue(), Dataset.AUDIT.getKey());
+    }
+
+    @Test
+    public void testEventMetadataAsList() {
+        PubSubToElasticsearchOptions options =
+                TestPipeline.testingPipelineOptions().as(PubSubToElasticsearchOptions.class);
+
+        options.setDeadletterTable("test:dataset.table");
+        options.setElasticsearchUsername("test");
+        options.setElasticsearchPassword("test");
+        options.setDataset(Dataset.AUDIT);
+        options.setNamespace("test-namespace");
+
+        String inputMessage = "{\"message\":\"test message\"}";
```

Comment:
Shouldn't missing timestamp property in input message throws NoSuchElementException

Reply:
@prathapreddy123 @suyograo
It's a good question. As discussed before, if timestamp is not present in input message we generate it on the fly.
But absence of timestamp can be pointing a bad message.